pkg/agent/monitoraddnodes.go

```diff
@@ -0,0 +1,272 @@
+			mon.logCSRsPendingApproval()
+			cancel()
+		}
+	}, 5*time.Second, waitContext.Done())
+
+	waitErr := waitContext.Err()
+	if waitErr != nil {
+		if errors.Is(waitErr, context.Canceled) {
+			cancel()
+		}
+		if errors.Is(waitErr, context.DeadlineExceeded) {
+			return errors.Wrap(waitErr, "monitor-add-nodes process timed out")
+		}
+	}
+
+	return nil
+}
+
+type addNodeMonitor struct {
+	nodeIPAddress string
+	cluster       *Cluster
+	currentState  addNodeState
+}
+
+func newAddNodeMonitor(nodeIP string, cluster *Cluster) *addNodeMonitor {
+	mon := addNodeMonitor{
+		nodeIPAddress: nodeIP,
+		cluster:       cluster,
+		currentState:  InitialState,
+	}
+	return &mon
+}
+
+func (mon *addNodeMonitor) setState(state addNodeState) {
+	mon.currentState = state
+}
+
+func (mon *addNodeMonitor) in(state addNodeState) bool {
+	return mon.currentState == state
+}
+
+func (mon *addNodeMonitor) onOrBefore(state addNodeState) bool {
+	return mon.currentState <= state
+}
+
+func (mon *addNodeMonitor) onOrAfter(state addNodeState) bool {
+	return mon.currentState >= state
+}
+
+func (mon *addNodeMonitor) currentStateName() string {
+	return stateNameMap[int(mon.currentState)]
+}
+
+func (mon *addNodeMonitor) nodeHasJoinedClusterAndIsReady() (bool, bool, error) {
+	nodes, err := mon.cluster.API.Kube.ListNodes()
+	if err != nil {
+		logrus.Debugf("error getting node list %v", err)
+		return false, false, nil
+	}
+
+	var joinedNode corev1.Node
+	hasJoined := false
+	for _, node := range nodes.Items {
+		for _, address := range node.Status.Addresses {
+			if address.Type == corev1.NodeInternalIP {
+				if address.Address == mon.nodeIPAddress {
+					joinedNode = node
+					hasJoined = true
+				}
+			}
+		}
+	}
+
+	isReady := false
+	if hasJoined {
+		logrus.Debugf("Node %v (%s) has joined cluster", mon.nodeIPAddress, joinedNode.Name)
+		for _, cond := range joinedNode.Status.Conditions {
+			if cond.Type == corev1.NodeReady && cond.Status == corev1.ConditionTrue {
+				isReady = true
+			}
+		}
+		if isReady {
+			logrus.Debugf("Node %s (%s) is Ready", mon.nodeIPAddress, joinedNode.Name)
+		} else {
+			logrus.Debugf("Node %s (%s) is not Ready", mon.nodeIPAddress, joinedNode.Name)
+		}
+	} else {
+		logrus.Debugf("Node %s has not joined cluster", mon.nodeIPAddress)
+	}
+
+	return hasJoined, isReady, nil
+}
+
+func (mon *addNodeMonitor) logCSRsPendingApproval() {
+	csrs, err := mon.cluster.API.Kube.ListCSRs()
+	if err != nil {
+		logrus.Debugf("error calling listCSRs(): %v ", err)
+	}
+
+	for _, csr := range csrs.Items {
+		if len(csr.Status.Conditions) == 0 {
+			// CSR is Pending and awaiting approval
+			logrus.Infof("CSR %s with signerName %s and username %s is Pending and awaiting approval",
+				csr.Name, csr.Spec.SignerName, csr.Spec.Username)
+		}
+	}
+}
+
+func (mon *addNodeMonitor) clusterHasFirstCSRPending() bool {
+	return len(mon.cluster.API.Kube.getCSRsPendingApproval("kubernetes.io/kube-apiserver-client-kubelet")) > 0
```

Comment:
Assuming that a multiple nodes installation is in progress, isn't this kind of check too loose? Can't be use ie the node name to verify that the csr is for that particular node?

Reply:
The first CSR doesn't contain a node name or IP address to identify which node it is for. 

The second CSR has the node name, but it is not clear to me how to find its IP address.
* Can we assume that the nodes-config.yaml file always contain the hostnames for the nodes being added?
* If not, how do we know which DNS server to query to check its IP address? 
* If the API_VIP is defined, can we assume it to be able to resolve hostnames?

Reply:
1. Yes, the hostname must be always specified. 

Not sure if there's a better way, but using `oc describe` it's possible to extract some information from the first csr (the `node-bootstrapper` one), for example: 

```
$ oc describe csr csr-4gflr | grep "Common Name"
         Common Name:    system:node:extra-worker-0
```
These IINW are retrieved by the stored certificate.

For the second csr, we could rely on the username field:
```
$ oc get csr csr-srhpw -o jsonpath='{.spec.username}'
system:node:extra-worker-0
```

Reply:
That's helpful. I will improve the CSR matching in another patch.